scene/resources/curve.cpp

```diff
@@ -1496,76 +1499,135 @@ void Curve3D::_bake() const {
+	// Step 2: Calculate the up vectors and the whole local reference frame
+	//
+	// See Dougan, Carl. "The parallel transport frame." Game Programming Gems 2 (2001): 215-219.
+	// for an example discussing about why not the Frenet frame.
+	{
+		PackedVector3Array forward_vectors;
 
-	for (const Plane &E : pointlist) {
-		w[idx] = E.normal;
-		wt[idx] = E.d;
-		wd[idx] = distlist[idx];
+		baked_up_vector_cache.resize(point_count);
+		forward_vectors.resize(point_count);
 
-		if (!up_vector_enabled) {
-			idx++;
-			continue;
+		Vector3 *up_write = baked_up_vector_cache.ptrw();
+		Vector3 *forward_write = forward_vectors.ptrw();
+
+		const Vector3 *points_ptr = baked_point_cache.ptr();
+
+		Basis frame; // X-right, Y-up, Z-forward.
+		Basis frame_prev;
+
+		// Set the initial frame based on Y-up rule.
+		{
+			Vector3 up(0, 1, 0);
+			Vector3 forward = (points_ptr[1] - points_ptr[0]).normalized();
+			if (forward.is_equal_approx(Vector3())) {
+				forward = Vector3(1, 0, 0);
+			}
+
+			if (abs(forward.dot(up)) > 1.0 - UNIT_EPSILON) {
+				frame_prev = Basis::looking_at(-forward, up);
+			} else {
+				frame_prev = Basis::looking_at(-forward, Vector3(1, 0, 0));
+			}
+
+			up_write[0] = frame_prev.get_column(1);
+			forward_write[0] = frame_prev.get_column(2);
 		}
 
-		forward = idx > 0 ? (w[idx] - w[idx - 1]).normalized() : prev_forward;
+		// Calculate the Parallel Transport Frame.
+		for (int idx = 1; idx < point_count; idx++) {
+			Vector3 forward = (points_ptr[idx] - points_ptr[idx - 1]).normalized();
+			if (forward.is_equal_approx(Vector3())) {
+				forward = frame_prev.get_column(2);
+			}
 
-		real_t y_dot = prev_up.dot(forward);
+			Basis rotate;
+			rotate.rotate_to_align(frame_prev.get_column(2), forward);
+			frame = rotate * frame_prev;
+			frame.orthonormalize(); // guard against float error accumulation
 
-		if (y_dot > (1.0f - CMP_EPSILON)) {
-			sideways = prev_sideways;
-			up = -prev_forward;
-		} else if (y_dot < -(1.0f - CMP_EPSILON)) {
-			sideways = prev_sideways;
-			up = prev_forward;
-		} else {
-			sideways = prev_up.cross(forward).normalized();
-			up = forward.cross(sideways).normalized();
+			up_write[idx] = frame.get_column(1);
+			forward_write[idx] = frame.get_column(2);
+
+			frame_prev = frame;
 		}
 
-		if (idx == 1) {
-			up_write[0] = up;
+		bool is_loop = true;
+		// Loop smoothing only applies when the curve is a loop, which means two ends meet, and share forward directions.
+		{
+			if (!points_ptr[0].is_equal_approx(points_ptr[point_count - 1])) {
+				is_loop = false;
+			}
+
+			real_t dot = forward_write[0].dot(forward_write[point_count - 1]);
+			if (dot < 1.0 - 0.01) { // Alignment should not be too tight, or it dosen't work for coarse bake interval
+				is_loop = false;
+			}
 		}
 
-		up_write[idx] = up;
+		// Twist up vectors, so that they align at two ends of the curve.
+		if (is_loop) {
+			const Vector3 up_start = up_write[0];
+			const Vector3 up_end = up_write[point_count - 1];
```

Comment:
It seems that the rotation still does not perfectly match with Progress 0 and 1.

![image](https://user-images.githubusercontent.com/61938263/201465077-59ec57d8-c568-42d1-b333-0bea51c06eed.png)

Possibly I guess the main cause is that the interval at the end of the bake is different from the other intervals. Yeah, now behavior is better than before, so I think it is fine for now, but there is room for improvement.

Reply:
Yes, I agree. I will investigate later.